### ISSUE TITLE
[Tests] recvmsg/sendto fails with "bad file descriptor"

### ISSUE THREAD

**Issue by hukoyu** [lsds/sgx-lkl#449]
2020-06-11T19:06:41.000000Z [(H)ERROR] tid(0x7f9f23fff700) | :OE_BUFFER_TOO_SMALL [/agent/_work/1/s/openenclave/host/sgx/sgxquote.c:oe_sgx_get_supported_attester_format_ids:109]
Azure Quote Provider: libdcap_quoteprov.so [ERROR]: Could not retreive environment variable for 'AZDCAP_DEBUG_LOG_LEVEL'

```
[[  SGX-LKL ]] sgxlkl_enclave_init(): enter
[[  SGX-LKL ]] sgxlkl_enclave_show_attribute(): enclave base=0x7f9e80000000 size=1.043 GB
[[  SGX-LKL ]] sgxlkl_enclave_show_attribute(): enclave heap base=0x7f9e80b5c000 size=1024.00 M end=0x7f9ec0b5c000
[[  SGX-LKL ]] sgxlkl_enclave_init(): calling _dlstart_c()
[[  SGX-LKL ]] _register_enclave_signal_handlers(): Registering OE exception handler...
[[  SGX-LKL ]] lkl_start_init(): kernel command line: 'mem=512m console=hvc0'
[[  SGX-LKL ]] lkl_start_init(): lkl_start_kernel() called
[    0.000000] Linux version 5.3.0+ (AzDevOps@cisca24ba000129) (gcc version 7.5.0 (Ubuntu 7.5.0-3ubuntu1~18.04)) #1 Thu Jun 11 18:46:42 UTC 2020
[    0.000000] memblock address range: 0x7f9ea08db000 - 0x7f9ec08db000
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 129280
[    0.000000] Kernel command line: mem=512m console=hvc0 virtio_mmio.device=256@0x1000000:1
[    0.000000] Dentry cache hash table entries: 65536 (order: 7, 524288 bytes, linear)
[    0.000000] Inode-cache hash table entries: 32768 (order: 6, 262144 bytes, linear)
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory available: 516256k/524288k RAM
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 4096
[    0.000000] lkl: irqs initialized
[    0.000000] clocksource: lkl: mask: 0xffffffffffffffff max_cycles: 0x1cd42e4dffb, max_idle_ns: 881590591483 ns
[    0.000000] lkl: time and timers initialized (irq2)
[    0.001617] printk: console [hvc0] enabled
[    0.001617] pid_max: default: 4096 minimum: 301
[    0.002124] Mount-cache hash table entries: 1024 (order: 1, 8192 bytes, linear)
[    0.002124] Mountpoint-cache hash table entries: 1024 (order: 1, 8192 bytes, linear)
[    0.005885] devtmpfs: initialized
[    0.013600] random: get_random_bytes called from 0x00007f9e80676515 with crng_init=0
[    0.014144] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.014144] futex hash table entries: 16 (order: -4, 384 bytes, linear)
[    0.014144] xor: automatically using best checksumming function   8regs     
[    0.014144] NET: Registered protocol family 16
[    0.019704] clocksource: Switched to clocksource lkl
[    0.019704] NET: Registered protocol family 2
[    0.020260] tcp_listen_portaddr_hash hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.020815] TCP established hash table entries: 4096 (order: 3, 32768 bytes, linear)
[    0.020815] TCP bind hash table entries: 4096 (order: 3, 32768 bytes, linear)
[    0.020815] TCP: Hash tables configured (established 4096 bind 4096)
[    0.021377] UDP hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.021377] UDP-Lite hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.021934] NET: Registered protocol family 1
[    0.021934] virtio-mmio: Registering device virtio-mmio.0 at 0x1000000-0x10000ff, IRQ 1.
[    0.021934] workingset: timestamp_bits=62 max_order=18 bucket_order=0
[    0.024684] NET: Registered protocol family 38
[    0.024684] io scheduler mq-deadline registered
[    0.024684] io scheduler kyber registered
[    0.160085] software IO TLB: mapped [mem 0x7f9f2eb97000-0x7f9f32b97000] (64MB)
[    0.162179] loop: module loaded
[    0.162179] device-mapper: ioctl: 4.40.0-ioctl (2019-01-18) initialised: dm-devel@redhat.com
[    0.164578] wireguard: WireGuard 0.0.20191219 loaded. See www.wireguard.com for information.
[    0.165136] wireguard: Copyright (C) 2015-2019 Jason A. Donenfeld <Jason@zx2c4.com>. All Rights Reserved.
[    0.165692] NET: Registered protocol family 10
[    0.166249] Segment Routing with IPv6
[    0.166804] sit: IPv6, IPv4 and MPLS over IPv4 tunneling driver
[    0.166804] This architecture does not have kernel memory protection.
[    0.175671] Run /init as init process
[[  SGX-LKL ]] lkl_start_init(): lkl_start_kernel() finished
[[  SGX-LKL ]] lkl_start_init(): creating LKL termination thread
[[  SGX-LKL ]] create_lkl_termination_thread(): enter
[[  SGX-LKL ]] lkl_start_init(): calling lkl_mount_virtial()
[[  SGX-LKL ]] lkl_termination_thread(): enter
[[  SGX-LKL ]] lkl_termination_thread(): Performed LKL syscall to get host task allocated (pid=36)
[[  SGX-LKL ]] lkl_start_init(): calling init_random()
[[  SGX-LKL ]] init_random(): Adding entropy to entropy pool
[    0.177108] random: crng init done
[[  SGX-LKL ]] init_enclave_clock(): Setting enclave realtime clock
[    0.231901] virtio_blk virtio1: [vda] 3072000 512-byte logical blocks (1.57 GB/1.46 GiB)
2020-06-11T19:06:41.000000Z [(H)ERROR] tid(0x7f9f23fff700) | :OE_BUFFER_TOO_SMALL [/agent/_work/1/s/openenclave/host/sgx/sgxquote.c:oe_sgx_get_supported_attester_format_ids:109]
Azure Quote Provider: libdcap_quoteprov.so [ERROR]: Could not retreive environment variable for 'AZDCAP_DEBUG_LOG_LEVEL'
[[  SGX-LKL ]] lkl_mount_disk(): lkl_mount_disk(dev="/dev/vda", mnt="/mnt/vda", ro=0)
[    0.234708] EXT4-fs (vda): recovery complete
[    0.234708] EXT4-fs (vda): mounted filesystem with ordered data mode. Opts: 
[[  SGX-LKL ]] lkl_mount_disks(): Set working directory: /
[[  SGX-LKL ]] libc_start_main_stage2(): Calling app main: /ltp/testcases/kernel/syscalls/sendto/sendto01
sendto01    0  [1;34mTINFO[0m  :  Thread created
sendto01    1  [1;32mTPASS[0m  :  bad file descriptor successful
[[  SGX-LKL ]] lkl_terminate(): terminating LKL (exit_status=1)
[[  SGX-LKL ]] lkl_termination_thread(): termination thread unblocked
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_sys_chdir(/)
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_virtio_netdev_remove()
[[  SGX-LKL ]] lkl_termination_thread(): calling lkl_sys_halt()
[    0.252325] reboot: Restarting system
[[  SGX-LKL ]] lkl_termination_thread(): lthread_detach2() done
```

**Comment by hukoyu** [comment]
Several LTP tests failed with this new error. This is a regression.
 
kernel-syscalls-sendto-sendto01-(debug)-(run-hw)-(8-ethreads)
kernel-syscalls-fchmodat-fchmodat01-(nonrelease)-(run-hw)-(8-ethreads)
kernel-syscalls-pwrite-pwrite04-(debug)-(run-hw)-(8-ethreads)

cc @letmaik @davidchisnall @prp @mikbras @jxyang @paulcallen

**Comment by letmaik** [comment]
Please report `Azure Quote Provider: libdcap_quoteprov.so [ERROR]: Could not retreive environment variable for 'AZDCAP_DEBUG_LOG_LEVEL'` [upstream](https://github.com/microsoft/Azure-DCAP-Client). This is only a warning and shouldn't really be printed as it's misleading.

**Comment by letmaik** [comment]
@hukoyu Your test is failing because of "bad file descriptor successful". This is unrelated to the two things you mention in the title of this issue. Could you separate things to avoid confusion?

**Comment by wintersteiger** [comment]
Yes, those warnings are spurious, you can ignore them. The `OE_BUFFER_TO_SMALL` is reported while computing the size of a buffer inside OE and the Azure DCAP driver reports that `AZDCAP_DEBUG_LOG_LEVEL`, isn't set, but it doesn't actually need it. I was planning to file an issue/submit a PR to OE for these, so they'll likely go away soon.

**Comment by hukoyu** [comment]
ltp2 (nonrelease, hw, 4 ethreads) / kernel-syscalls-recvmsg-recvmsg01-(nonrelease)-(run-hw)-(4-ethreads) failed with same error:
```
[    0.000000] Linux version 5.3.0+ (AzDevOps@cisca24ba00014M) (gcc version 7.5.0 (Ubuntu 7.5.0-3ubuntu1~18.04)) #1 Mon Jun 15 00:04:54 UTC 2020
[    0.000000] memblock address range: 0x7ffa207f5000 - 0x7ffa407f5000
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 129280
[    0.000000] Kernel command line: mem=512m console=hvc0 virtio_mmio.device=256@0x1000000:1
[    0.000000] Dentry cache hash table entries: 65536 (order: 7, 524288 bytes, linear)
[    0.000000] Inode-cache hash table entries: 32768 (order: 6, 262144 bytes, linear)
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] Memory available: 516256k/524288k RAM
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 4096
[    0.000000] lkl: irqs initialized
[    0.000000] clocksource: lkl: mask: 0xffffffffffffffff max_cycles: 0x1cd42e4dffb, max_idle_ns: 881590591483 ns
[    0.000000] lkl: time and timers initialized (irq2)
[    0.000000] printk: console [hvc0] enabled
[    0.000502] pid_max: default: 4096 minimum: 301
[    0.000503] Mount-cache hash table entries: 1024 (order: 1, 8192 bytes, linear)
[    0.000503] Mountpoint-cache hash table entries: 1024 (order: 1, 8192 bytes, linear)
[    0.001565] devtmpfs: initialized
[    0.004905] random: get_random_bytes called from 0x00007ffa0059dd81 with crng_init=0
[    0.004905] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.005462] futex hash table entries: 16 (order: -4, 384 bytes, linear)
[    0.005462] xor: automatically using best checksumming function   8regs     
[    0.005462] NET: Registered protocol family 16
[    0.008915] clocksource: Switched to clocksource lkl
[    0.009476] NET: Registered protocol family 2
[    0.010034] tcp_listen_portaddr_hash hash table entries: 256 (order: 0, 4096 bytes, linear)
[    0.010034] TCP established hash table entries: 4096 (order: 3, 32768 bytes, linear)
[    0.010034] TCP bind hash table entries: 4096 (order: 3, 32768 bytes, linear)
[    0.010034] TCP: Hash tables configured (established 4096 bind 4096)
[    0.010034] UDP hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.010034] UDP-Lite hash table entries: 256 (order: 1, 8192 bytes, linear)
[    0.010034] NET: Registered protocol family 1
[    0.010590] virtio-mmio: Registering device virtio-mmio.0 at 0x1000000-0x10000ff, IRQ 1.
[    0.010590] workingset: timestamp_bits=62 max_order=18 bucket_order=0
[    0.011708] NET: Registered protocol family 38
[    0.011708] io scheduler mq-deadline registered
[    0.011708] io scheduler kyber registered
[    0.118978] software IO TLB: mapped [mem 0x7ffa8820d000-0x7ffa8c20d000] (64MB)
[    0.120652] loop: module loaded
[    0.123992] device-mapper: ioctl: 4.40.0-ioctl (2019-01-18) initialised: dm-devel@redhat.com
[    0.125103] wireguard: WireGuard 0.0.20191219 loaded. See www.wireguard.com for information.
[    0.125103] wireguard: Copyright (C) 2015-2019 Jason A. Donenfeld <Jason@zx2c4.com>. All Rights Reserved.
[    0.125658] NET: Registered protocol family 10
[    0.128912] Segment Routing with IPv6
[    0.128912] sit: IPv6, IPv4 and MPLS over IPv4 tunneling driver
[    0.129419] This architecture does not have kernel memory protection.
[    0.132188] Run /init as init process
[    0.132752] random: crng init done
[    0.169120] virtio_blk virtio1: [vda] 3072000 512-byte logical blocks (1.57 GB/1.46 GiB)
[    0.170790] EXT4-fs (vda): mounted filesystem with ordered data mode. Opts: 
recvmsg01    1  [1;32mTPASS[0m  :  bad file descriptor successful
recvmsg01    2  [1;32mTPASS[0m  :  invalid socket successful
recvmsg01    3  [1;32mTPASS[0m  :  invalid socket buffer successful
recvmsg01    4  [1;32mTPASS[0m  :  invalid socket length successful
recvmsg01    5  [1;32mTPASS[0m  :  invalid iovec count successful
recvmsg01    6  [1;32mTPASS[0m  :  rights reception successful
recvmsg01    7  [1;32mTPASS[0m  :  invalid MSG_OOB flag set successful
recvmsg01    8  [1;32mTPASS[0m  :  invalid MSG_ERRQUEUE flag set successful
recvmsg01    9  [1;32mTPASS[0m  :  invalid cmsg length successful
recvmsg01   10  [1;32mTPASS[0m  :  large cmesg length successful
server select: Bad file descriptor
[    0.181140] EXT4-fs (vda): re-mounted. Opts: (null)
[    0.181140] reboot: Restarting system
2020-06-15T00:23:34.000000Z [(H)ERROR] tid(0x7ffa84a01700) | :OE_BUFFER_TOO_SMALL [/agent/_work/1/s/openenclave/host/sgx/sgxquote.c:oe_sgx_get_supported_attester_format_ids:109]
Azure Quote Provider: libdcap_quoteprov.so [ERROR]: Could not retreive environment variable for 'AZDCAP_DEBUG_LOG_LEVEL'
```

**Comment by wintersteiger** [comment]
A fix for the spurious `OE_BUFFER_TOO_SMALL` was merged into OE master earlier today: https://github.com/openenclave/openenclave/pull/3129. @jhand2 could you rebase `feature/sgx-lkl-support` again? Thanks!

The bad file descriptor is likely a real problem, but I have no idea what that file descriptor is for and whether some of the LTP patches that seem to be close to those error messages make a difference here.

**Comment by SeanTAllen** [comment]
It's unclear to me what the resolution of this was that led to this being closed.

 What was the issue with the bad file descriptor message? 
Was that the result of some code not correctly handling shutdown?

**Comment by prp** [comment]
@hukoyu since you're re-opened this, can you clean up the issue? Output like this is not sufficient to debug the issue:
```
recvmsg01    1  [1;32mTPASS[0m  :  bad file descriptor successful
recvmsg01    2  [1;32mTPASS[0m  :  invalid socket successful
recvmsg01    3  [1;32mTPASS[0m  :  invalid socket buffer successful
recvmsg01    4  [1;32mTPASS[0m  :  invalid socket length successful
recvmsg01    5  [1;32mTPASS[0m  :  invalid iovec count successful
recvmsg01    6  [1;32mTPASS[0m  :  rights reception successful
recvmsg01    7  [1;32mTPASS[0m  :  invalid MSG_OOB flag set successful
recvmsg01    8  [1;32mTPASS[0m  :  invalid MSG_ERRQUEUE flag set successful
recvmsg01    9  [1;32mTPASS[0m  :  invalid cmsg length successful
recvmsg01   10  [1;32mTPASS[0m  :  large cmesg length successful
server select: Bad file descriptor
```
Can you:
- include details what the test is doing?
- describe what exactly is failing?
- list the exact call arguments that don't work? (You can use the SGX-LKL syscall tracing features to get more detail.)
- update the issue title to be more descriptive?

We can then triage this. Thanks.

**Comment by prp** [comment]
@davidchisnall this keeps failing the `recvmsg/sendto` LTP tests in a non-deterministic fashion. Do we have anybody who can look into this? It may actually be a problem with the test itself -- I'm not sure.

**Comment by paulcallen** [comment]
@hukoyu @Pengpeng-Microsoft?

**Comment by hukoyu** [comment]
recvmsg fixed with PR https://github.com/lsds/sgx-lkl/pull/578
sendto will be fixed with another PR as well.

**Comment by hukoyu** [comment]
Checking also if we have same issue with these patches as well:
recvfrom01
sendto
send01
sendmsg01
sendfile02,05,06,09

**Comment by hukoyu** [comment]
recvmsg fixed with PR https://github.com/lsds/sgx-lkl/pull/578
send01 fixed with PR https://github.com/lsds/sgx-lkl/pull/586
sendto01 fixed with PR https://github.com/lsds/sgx-lkl/pull/587
recvfrom01 fixed with PR https://github.com/lsds/sgx-lkl/pull/583
sendfile02 fixed with PR https://github.com/lsds/sgx-lkl/pull/588
sendmsg01 fixed with PR https://github.com/lsds/sgx-lkl/pull/585

sendfile 05,06, 09 left. All others fixed.

**Comment by hukoyu** [comment]
https://github.com/lsds/sgx-lkl/pull/595 fixed the sendfile 05,06,09 tests. 
All of the enabled tests are addressed for bad file descriptor issue and fixed.

**Comment by hukoyu** [comment]
I did a simple grep in the patches and we also use close() for file descriptors in these syscalls. 
It doesn’t mean these patches have problem. We did a quick code review for the patches and determined if we have bad file descriptor issue for these patches or not. Total 21 tests listed below. Let’s first figure out which ones needs to be fixed and then submit PRs to fix them. 

-    Fcntl24,25,32
-    Connect01
-    Epoll-ltp
-    Pipe02,04,09
-    setresuid04
-    setreuid07
-    getxattr01,02,03
-    fchmod03_fchmod04_fchown03
-    fstatfs02
-    open08
-    renameat201,202
-    setfsuid04

Result of code review:
Only connect01 and setreuid07 had bad file descriptor issue and fixed with PRs https://github.com/lsds/sgx-lkl/pull/593 and https://github.com/lsds/sgx-lkl/pull/590 
All other tests are OK. No fix needed.

**Comment by hukoyu** [comment]
We addressed bad file descriptor for recvmsg, sendto and all other enabled tests. All tests are reviewed from this perspective and fixed.
